v8env/src/fly/fetch/proxy.ts

```diff
@@ -0,0 +1,159 @@
+ * // sends all traffic to an Amazon ELB,
+ * // `Host` header passes through from visitor request
+ * const origin = proxy("https://elb1298.amazonaws.com")
+ * ```
+ * 
+ * By default, this function sends the `Host` header inferred from the origin URL. To forward
+ * host headers sent by visitors, set `forwardHostHeader` to true.
+ * 
+ * ```javascript
+ * // sends all traffic to an Amazon ELB, include host header from original request.
+ * const origin = proxy("https://elb1298.amazonaws.com", {
+ *  forwardHostHeader: true
+ * })
+ * ```
+ * 
+ * And then way more rare, no host header at all. Usually you'd strip out `x-forwarded-host`, 
+ * since some origins don't like that:
+ * ```javascript
+ * // sends all traffic to an Amazon ELB, never sends a host header
+ * const origin = proxy("https://elb1298.amazonaws.com", {
+ *  headers: { host: false}
+ * })
+ * ```
+ * 
+ * @preferred
+ * @module fly/fetch/proxy
+ */
+
+/**
+ * This generates a `fetch` like function for proxying requests to a given origin.
+ * When this function makes origin requests, it adds standard proxy headers like 
+ * `X-Forwarded-Host` and `X-Forwarded-For`. It also passes headers from the original
+ * request to the origin.
+ * @param origin A URL to an origin, can include a path to rebase requests.
+ * @param options Options and headers to control origin request.
+ */
+export function proxy(origin: string | URL, options?: ProxyOptions) {
+  return function proxyFetch(req: RequestInfo, init?: RequestInit) {
+    if (!options) {
+      options = {}
+    }
+    const breq = buildProxyRequest(origin, options, req, init)
+    return fetch(breq)
+  }
+}
+
+export default proxy
+
+interface FlyRequest extends Request {
+  url: string
+}
+
+/**
+ * @protected
+ * @hidden
+ * @param origin 
+ * @param options 
+ * @param req 
+ * @param init 
+ */
+export function buildProxyRequest(origin: string | URL, options: ProxyOptions, req: RequestInfo, init?: RequestInit) {
+
+  if (typeof req === "string") {
+    req = new Request(req)
+  }
+  const url = new URL(req.url)
+  let breq: FlyRequest | null = null
+
+  if (req instanceof Request) {
+    breq = req.clone()
+  } else {
+    breq = new Request(req)
+  }
+
+  if (typeof origin === "string") {
+    origin = new URL(origin)
+  }
+
+  url.hostname = origin.hostname
+  url.protocol = origin.protocol
+  url.port = origin.port
+
+  if (options.stripPath && typeof options.stripPath === 'string') {
+    // remove basePath so we can serve `onehosthame.com/dir/` from `origin.com/`
+    url.pathname = url.pathname.substring(options.stripPath.length)
+  }
+  if (origin.pathname && origin.pathname.length > 0) {
+    url.pathname = [origin.pathname.replace(/\/$/, ''), url.pathname.replace(/^\//, "")].join("/")
+  }
+  if (url.pathname.startsWith("//")) {
+    url.pathname = url.pathname.substring(1)
+  }
+
+  breq.url = url.toString()
+  // we extend req with remoteAddr
+  breq.headers.set("x-forwarded-for", (<any>req).remoteAddr)
```

Comment:
Don't need the any cast anymore 😄 https://github.com/superfly/fly/blob/master/v8env/src/request.ts#L31